lib/active_merchant/billing/gateways/hps.rb

```diff
@@ -254,27 +258,27 @@ def add_descriptor_name(xml, options)
         xml.hps :TxnDescriptor, options[:descriptor_name] if options[:descriptor_name]
       end
 
-      def add_three_d_secure(xml, card_or_token, options)
-        if card_or_token.is_a?(NetworkTokenizationCreditCard)
-          build_three_d_secure(xml, {
-            source: card_or_token.source,
-            cavv: card_or_token.payment_cryptogram,
-            eci: card_or_token.eci,
-            xid: card_or_token.transaction_id
-          })
-        elsif options[:three_d_secure]
-          options[:three_d_secure][:source] ||= card_brand(card_or_token)
-          build_three_d_secure(xml, options[:three_d_secure])
+      def add_wallet_data(xml, payment_method, options)
+        return unless payment_method.is_a?(NetworkTokenizationCreditCard)
+
+        xml.hps :WalletData do
+          xml.hps :PaymentSource, PAYMENT_DATA_SOURCE_MAPPING[payment_method.source]
+          xml.hps :Cryptogram, payment_method.payment_cryptogram
+          xml.hps :ECI, payment_method.eci&.to_s&.sub!(/^0/, '') if payment_method.eci
         end
       end
 
-      def build_three_d_secure(xml, three_d_secure)
+      def add_three_d_secure(xml, card_or_token, options)
         # PaymentDataSource is required when supplying the SecureECommerce data group,
         # and the gateway currently only allows the values within the mapping
-        return unless PAYMENT_DATA_SOURCE_MAPPING[three_d_secure[:source].to_sym]
+        return unless (three_d_secure = options[:three_d_secure])
+
+        three_d_secure[:source] ||= card_brand(card_or_token)
+
+        return unless (source = PAYMENT_DATA_SOURCE_MAPPING[three_d_secure[:source].to_sym])
 
         xml.hps :SecureECommerce do
```

Comment:
I think this line needs to be the `:Secure3D` now according to the docs in the ticket.

Reply:
Or will there be a separate ticket for the 3dsecure updates?

Reply:
Ah thanks for pointing that out, I definitely missed that bullet point.

Reply:
Even though we have 3DS fields here it doesn't look like we support 3DS on our side for this gateway